temporalio/worker/_workflow.py

```diff
@@ -421,3 +439,97 @@ def nondeterminism_as_workflow_fail_for_types(self) -> Set[str]:
                 for typ in v.failure_exception_types
             )
         )
+
+
+class _DeadlockError(Exception):
+    """Exception class for deadlocks. Contains functionality to swap the default traceback for another."""
+
+    def __init__(self, message: str, replacement_tb: Optional[TracebackType] = None):
+        """Create a new DeadlockError, with message `msg` and optionally a traceback `tb` to be swapped in later.
+
+        Args:
+            message: Message to be presented through exception.
+            replacement_tb: Optional TracebackType to be swapped later.
+        """
+        super().__init__(message)
+        self._new_tb = replacement_tb
+
+    def has_pending_tb(self) -> bool:
+        """Check if the DeadlockError has a pending traceback. For use with swap_traceback().
+
+        Returns:
+            True if there is a pending traceback, false if not.
+        """
+        if self._new_tb:
+            return True
+        return False
+
+    def swap_traceback(self) -> None:
+        """Swap the current traceback for the replacement passed during construction. Used to work around Python adding the current frame to the stack trace.
+
+        Returns:
+            None
+        """
+        if self._new_tb:
+            self.__traceback__ = self._new_tb
+            self._new_tb = None
+
+    @classmethod
+    def from_deadlocked_workflow(
+        cls, workflow: WorkflowInstance, timeout: Optional[int]
+    ):
+        msg = f"[TMPRL1101] Potential deadlock detected: workflow didn't yield within {timeout} second(s)."
+        tid = workflow.get_thread_id()
+        if not tid:
+            return cls(msg)
+
+        tb, warn = cls._gen_tb_helper(tid)
```

Comment:
I think you need a try/except around this, and deadlocks are such a rare thing you don't need advanced warn machinations. Something like:

```suggestion
        try:
            tb = cls._gen_tb_helper(tid)
            if tb:
                return cls(msg, tb)
            return cls(f"{msg} (no frames available)")
        except Exception as err:
            return cls(f"{msg} (failed getting frames: {err})")
```

would be just fine. No need for advanced warn system for the different ways that getting trace can fail, and you get the benefit of trapping _all_ errors trying to collect traceback. But this is not a big deal.